pageserver/src/tenant.rs

```diff
@@ -1389,13 +1444,106 @@ impl Tenant {
+                rx
+            } else {
+                // try another time
+                let (tx, rx) = tokio::sync::broadcast::channel(1);
+                // now anyone else racing will see the None
+                let this = self.clone();
+                let timeline = timeline.clone();
+
+                let rx = Arc::new(rx);
+
+                // sadly futures::future::FutureExt::shared requires clone, which we cannot give
+                // for tokio::task::JoinError
+                //
+                // TODO: this could be tenant scoped task_mgr task?
+                tokio::spawn({
+                    let rx = rx.clone();
+                    async move {
+                        // to uphold the MaybeDone promise, we keep the channel alive *until* we've
+                        // swapped the values
+                        let tx = tx;
+                        let rx = rx;
+
+                        let res = this.unique_delete_timeline(&timeline).await;
+
+                        {
+                            let mut g = timeline.delete_self.lock().await;
+                            #[cfg(debug_assertions)]
+                            match g.as_ref() {
+                                Some(timeline::MaybeDone::Pending(weak)) => {
+                                    let same = weak
+                                        .upgrade()
+                                        // we don't yet have Reciver::same_channel
+                                        .map(|rx2| Arc::ptr_eq(&rx, &rx2))
+                                        .unwrap_or(false);
+                                    assert!(same, "different channel had been replaced");
+                                }
+                                other => panic!("unexpected MaybeDone: {other:?}"),
+                            }
+                            *g = Some(timeline::MaybeDone::Done(res.clone()));
+                            drop(rx);
+                        }
+
+                        // now no one can get the Pending(weak) value to upgrade and they only see
+                        // the Done(res).
+
+                        // send the result value to listeners, if any
+                        drop(tx.send(res));
+                    }
+                    .instrument(tracing::info_span!("unique_delete_timeline"))
+                });
+                *g = Some(timeline::MaybeDone::Pending(Arc::downgrade(&rx)));
+                rx
+            }
+        };
+
+        {
+            use tokio::sync::broadcast::error::RecvError;
+
+            match rx.resubscribe().recv().await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(e)) => Err(DeleteTimelineError::from(&e)),
+                // lagged doesn't mean anything with 1 send, but whatever, handle it the same
+                Err(RecvError::Closed | RecvError::Lagged(_)) => {
+                    Err(DeleteTimelineError::InternalFailure)
+                }
```

Comment:
I think we might still just see the `Closed`, if very unlucky, since the resubscribe is being done *outside* of the lock being held.

Reply:
fa947f0f